File: src/07_Advanced_Pre_DSA/STL_vs_Collections.java
/**
 * Level 7: Advanced Pre-DSA & Standard Libraries
 * File: STL_vs_Collections.java
 * Topic: Java Collections (ArrayList, HashMap, PriorityQueue, Collections.sort with Lambda)
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.PriorityQueue;

public class STL_vs_Collections {
    public static void main(String[] args) {
        System.out.println("--- 1. ArrayList & Collections.sort with Custom Lambda ---");
        ArrayList<Integer> nums = new ArrayList<>();
        nums.add(40); nums.add(10); nums.add(30); nums.add(20);

        // Sort descending using lambda comparator
        Collections.sort(nums, (a, b) -> b - a);

        for (int x : nums) System.out.print(x + " ");
        System.out.println();

        System.out.println("\n--- 2. HashMap (O(1) Hash Table) ---");
        HashMap<String, Integer> freq = new HashMap<>();
        freq.put("apple", 3);
        freq.put("banana", 5);
        System.out.println("Apple count: " + freq.get("apple"));

        System.out.println("\n--- 3. PriorityQueue (Min-Heap Default) ---");
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        minHeap.add(10);
        minHeap.add(5);
        minHeap.add(20);

        System.out.println("Min-Heap peek: " + minHeap.peek()); // 5
    }
}

File: src/07_Advanced_Pre_DSA/STL_vs_Collections.cpp
/**
 * Level 7: Advanced Pre-DSA & Standard Libraries
 * File: STL_vs_Collections.cpp
 * Topic: C++ STL Containers (vector, map, priority_queue, sort + custom lambda comparator)
 */

#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>

int main() {
    std::cout << "--- 1. Vector & std::sort with Custom Lambda ---\n";
    std::vector<int> nums = {40, 10, 30, 20};
    
    // Sort descending using lambda comparator
    std::sort(nums.begin(), nums.end(), [](int a, int b) {
        return a > b;
    });

    for (int x : nums) std::cout << x << " ";
    std::cout << "\n";

    std::cout << "\n--- 2. Unordered Map (O(1) Hash Table) ---\n";
    std::map<std::string, int> freq;
    freq["apple"] = 3;
    freq["banana"] = 5;
    std::cout << "Apple count: " << freq["apple"] << "\n";

    std::cout << "\n--- 3. Priority Queue (Min-Heap) ---\n";
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;
    minHeap.push(10);
    minHeap.push(5);
    minHeap.push(20);

    std::cout << "Min-Heap top: " << minHeap.top() << "\n"; // 5

    return 0;
}
